### PR TITLE
WIP: update engine to store all metadata in tsm files

### PR DESCRIPTION
**Pull request by pauldix**
ok, as promised here's the work in progress. It actually compiles, but there's still more I have to do and  the tests don't run. Anyway, let me know what you think of the structure.

Here's the high level concept of what happened:

Previously, keys (series key + field) would be hashed to a `uint64` ID. Because of the possibility of hash collisions, we had to keep a map of the ID to key in a file called `ids`. There was another file called `collisions` that would keep the map of the keys that had IDs different than their hashes. The collisions would be kept in memory.

This ensured that the key -> ID mapping was unique. The data files had a structure like:

```
# whole file
[block1, block2, blockN... index]

# block
[id, block len, compressed block]

# index
[id1, starting position in file, id2, pos2, idN, posN, ...]
```

The index and the blocks in the file would always be kept in ID ascending order, which made it possible to merge files together quickly and to search in the index efficiently using binary sort.

This also meant that other files needed to be kept to store metadata.

This PR updates the structure so that tsm files are self describing. That is they have all the information about metadata in them. The new structure looks like this:

```
# whole file
[block1, block2, blockN... index]

# block
[key len, compressed key, block len, compressed block]

# index
[id1, starting position, min time, max time, id2, ...]
```

The blocks no longer have the ID in them. Instead the first block will have the entire key (compressed with Snappy). Since blocks have no more than 1,000 points by default, there can be successive blocks with the same key. Those blocks will have a key length of 0 indicating that they belong to the same key as the previous blocks.

I keep the ID based index so that we can still quickly do a binary search to find the starting position of a given ID. Keys get hashed the same as before, the difference is that hash collisions get handled after we seek to the block level.

If there are two keys that hash to the same ID, they get stored in the data file in sorted lexicographical order.

The min and max times of the given ID in the file enable more complex compaction and rewrite logic. Data files can now have overlapping time ranges, as long as they're increasing and two data files can't have an overlapping time range for a specific ID.

Finally, I added some structs to break the code out a bit more and get rid of the monster Write and Rewrite methods.

I'll be updating it to remove all the metadata files, which also removes the need to store those separately from the point writes.

The loading of the metadata index on startup will jump from block to block in the data files and read the keys, which contain all the metadata information we need (measurement, tags, and the fields). Field types are already encoded in the compressed blocks themselves.

### CODE REVIEW

**Comment by otoolep** [comment]
Any chance we can get some high-level design notes? A short outline of what
we are looking at?

**Comment by pauldix** [comment]
@otoolep updated the description

File: tsdb/engine/tsm1/wal.go [hunk 1]
```diff
@@ -285,6 +288,10 @@ func (l *Log) addToCache(points []models.Point, fields map[string]*tsdb.Measurem
 	for _, p := range points {
 		for name, value := range p.Fields() {
 			k := SeriesFieldKey(string(p.Key()), name)
+			if len(k) > maxKeySize {
```
Comment (jwilder) [inline tsdb/engine/tsm1/wal.go]: Why is there a `maxKeySize` now?

Comment (pauldix) [inline tsdb/engine/tsm1/wal.go]: because I have to encode its length and I use a `uint16` for it. 65k should be enough for anyone, no?

**Comment by jwilder** [comment]
https://github.com/influxdb/influxdb/blob/pd-fix-rewrite-file/tsdb/engine/tsm1/tsm1.go#L99-L100
Could use some comments as to what they protect.

**Comment by jwilder** [comment]
https://github.com/influxdb/influxdb/blob/pd-fix-rewrite-file/tsdb/engine/tsm1/tsm1.go#L427-L465

Seems like this func could be made more simple and easier to understand.  

Some suggestions:

1) Move the inner loop of `splitValuesByTime` to separate func (`splitByTime`) that takes only a `time` and a `[]*valuesWithKey` slice and returns the left and right `[]*valuesWithKey` items.
2) Have `splitByTime` loop over each items and either append to left or right.  Sort both at end (if they are supposed to be sorted).  The `sort.Search` inside two nested looks complicated and expensive, but without benchmarking it's hard to say.

**Comment by jwilder** [comment]
https://github.com/influxdb/influxdb/blob/pd-fix-rewrite-file/tsdb/engine/tsm1/tsm1.go#L594-L620

Using channels instead of a `Mutex` might be cleaner here.

**Comment by jwilder** [comment]
https://github.com/influxdb/influxdb/blob/pd-fix-rewrite-file/tsdb/engine/tsm1/tsm1.go#L662-L664

Is this missing a `min < lockStart` case?

**Comment by jwilder** [comment]
https://github.com/influxdb/influxdb/blob/pd-fix-rewrite-file/tsdb/engine/tsm1/tsm1.go#L1036

`rottate`

**Comment by jwilder** [comment]
`tsm1.go` is getting really big.  Should probably break it out into separate files like `engine.go`, `checkout point.go`, `rewrite.go`, etc..  soon.

**Comment by jwilder** [comment]
The hashing of keys still seems problematic to me.   Having to deal with collisions adds some complexity as well as extra IO for some series that just happens to get a collision.   The full file scan at load time will be problematic for large files as well.  

Since you know all the keys that are going to be written into the file beforehand, why not assign them an ID where creating the file?  A simple way might be to sort all the keys and use their index position as the ID?  This sorted key list would be stored in the file and read at startup instead of the whole file. This would remove all collisions and make startups fast regardless of the size of the file.

**Comment by pauldix** [comment]
@jwilder the reason the hashing is preferable is so that when a query comes in, we don't need to hit a lookup table to find the ID for the given key.

Otherwise, we'd need to keep that map in memory for every file in the DB, which would blow up our memory size.

Unless there's some good way to have the keys in the file and be able to search through them efficiently?

**Comment by otoolep** [comment]
I think we can close this.

**Pull request closed without merging**